src/mlpack/core/data/imputation_methods/median_imputation.hpp

```diff
@@ -0,0 +1,64 @@
+/**
+ * @file median_imputation.hpp
+ * @author Keon Kim
+ *
+ * Definition and Implementation of the MedianImputation class.
+ */
+#ifndef MLPACK_CORE_DATA_IMPUTE_STRATEGIES_MEDIAN_IMPUTATION_HPP
+#define MLPACK_CORE_DATA_IMPUTE_STRATEGIES_MEDIAN_IMPUTATION_HPP
+
+#include <mlpack/core.hpp>
+
+using namespace std;
+
+namespace mlpack {
+namespace data {
+
+/**
+ * A simple median imputation
+ * replace missing value with middle or average of middle values
+ */
+template <typename T>
+class MedianImputation
+{
+ public:
+  void Impute(const arma::Mat<T>& input,
+              arma::Mat<T>& output,
+              const T& mappedValue,
+              const size_t dimension,
+              const bool transpose = true)
+  {
+    //initiate output
+    output = input;
+
+    if (transpose)
+    {
+      arma::Mat<T> medianMat = arma::median(input, 1);
+      for (size_t i = 0; i < input.n_cols; ++i)
+      {
+        if (input(dimension, i) == mappedValue ||
+            std::isnan(input(dimension, i)))
+        {
+          output(dimension, i) = medianMat(dimension, 0);
+        }
+      }
```

Comment:
This pattern seems to exist in all of the imputation strategies, so my comment here applies to all of them.  Here you are getting a single value out of each column, and each column is held contiguous in memory.  So each element you access is probably not in the same cache line, and as a result will probably be slower.  So if I were to call the imputer with every single dimension sequentially, it would be a lot slower than just taking one pass over the matrix and imputing each element as we went.  (It would also be slower in this case because we would be calculating the mean for every dimension every time we called the method.)